### ISSUE TITLE
Type of variable is inferred from literal instead of annotation

### ISSUE THREAD

**Issue by JakenVeina** [microsoft/TypeScript#40860]
**Search Terms:** 
variable
literal
type inference
type annotation
TS2367
TS2339

**Code**
```ts
type Foo = number | "test";

let x: Foo = 5;
let y = <Foo>5;

function bar() {
    x = "test";
    y = "test";
}
bar();

if (x === "test") { // TS2367 This condition will always return 'false' since the types 'number' and 'string' have no overlap.
    console.log(x.length); // TS2339 Property 'length' does not exist on type 'number';
}
if (y === "test") {
    console.log(y.length);
}
```

**Expected behavior:**
No error should be produced here, since the variable `x` was declared to be of type `Foo`, not `number`, even though it was initialized with a `number` literal, it's not declared as a `const` and thus could change (and in fact does).

**Actual behavior:**
The compiler seems to infer the type of `x` as `number`, from the literal `5` used to initialize it, even though the variable type is explicitly declared to be of type `Foo`.

**Playground Link:**
[Playground](https://www.typescriptlang.org/play?jsx=0&module=0&ts=4.0.2#code/C4TwDgpgBAYg9nKBeKA7ArgWwEYQE5QA+UARMBAM7AkDcAUHQDYTBQAeAXLAslAKz1mrELwA88OAD4BDAGbpUAY2ABLOKijYAhngAUASigBvOlDPteZStXrmoIlFaq06AXzra9++nRWyoumzISI7kzoZGUAD0UVAAKgDKAEwAzABsAOzxABYqFFCK6gAmKqrqUADuKoyMUFqMFVog+Xgs6HgaAOSy9RQQnVAUKkrQwNmj4JRQnRg4+ANaqEXTVHjDAOYD2VoAbtCoiHB7eIxaYAB0puaFqBRwzOeMcOuBjxCo62Pe0bGJqSkATigAAU8HBIHhQNNmB8xgMinApgdWBA2HlWOVQJBprNcHhOvR3H4Ag4QqQwtQIlczDc7g8ni8QG9YdlvG46EA)

**Related Issues:**
[#26979](https://github.com/microsoft/TypeScript/issues/26979)
[#10195](https://github.com/microsoft/TypeScript/issues/10195)

**Comment by RyanCavanaugh** [comment]
Your real complaint here is that the side effect in the function wasn't considered; see #9998.

Initialization is the same as assignment, so
```ts
let x: Foo = 5;
```
gets the same treatment as
```ts
let x: Foo = (whatever);
x = 5;
```
Given this, it would be ridiculous to error on
```ts
let x: Foo = (whatever);
x = 5;
x.toFixed(); // Error, Foo might be "test" ?????
```

**Comment by JakenVeina** [comment]
That's fair. I also think it's fair for the compiler to not recognize that side-effect. It can't investigate everything.

For context, the real-world scenario where this came up was...

```cs
interface ValidationErrorMap {
    readonly [errorCode: string]: string | ValidationErrorMap;
}

type ValidationResult = ValidationErrorMap | "success";

function combine(ReadonlyArray<ValidationResult>results): ValidationResult {
    let finalResult: ValidationResult = "success";

    for(const result of results) {
        finalResult = (result === "success") ? finalResult
            : (finalResult === "success") ? result
            : { ...finalResult, ...result };
    }

    return finalResult;
}
```

Here, I would definitely expect the side-effect to be recognizable, as it's happening in the same scope, just inside a loop.
